common/flatpak-run.c

```diff
@@ -3958,22 +4016,95 @@ flatpak_run_app (FlatpakDecomposed *app_ref,
 
   flatpak_bwrap_add_args (bwrap,
                           "--ro-bind", flatpak_file_get_path_cached (runtime_files), "/usr",
-                          "--lock-file", "/usr/.ref",
                           NULL);
 
+  if (runtime_files == original_runtime_files)
+    {
+      /* All true Flatpak runtimes have files/.ref */
+      flatpak_bwrap_add_args (bwrap,
+                              "--lock-file", "/usr/.ref",
+                              NULL);
+    }
+  else
+    {
+      g_autoptr(GFile) runtime_child = NULL;
+
+      runtime_child = g_file_get_child (runtime_files, ".ref");
+
+      /* Lock ${usr}/.ref if it exists */
+      if (g_file_query_exists (runtime_child, NULL))
+        flatpak_bwrap_add_args (bwrap,
+                                "--lock-file", "/usr/.ref",
+                                NULL);
+
+      /* Put the real Flatpak runtime in /run/parent, so that the
+       * replacement /usr can have symlinks into /run/parent in order
+       * to use the Flatpak runtime's graphics drivers etc. if desired */
+      flatpak_bwrap_add_args (bwrap,
+                              "--ro-bind",
+                              flatpak_file_get_path_cached (original_runtime_files),
+                              "/run/parent/usr",
+                              "--lock-file", "/run/parent/usr/.ref",
+                              NULL);
+      flatpak_run_setup_usr_links (bwrap, original_runtime_files,
+                                   "/run/parent");
+
+      g_clear_object (&runtime_child);
+      runtime_child = g_file_get_child (runtime_files, "etc");
```

Comment:
Shouldn't this be original_runtime_files?

Reply:
Yes, good catch.